packages/flutter/lib/src/rendering/object.dart

```diff
@@ -843,49 +848,79 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
     return _layer;
   }
 
-  bool _needsCompositingBitsUpdate = true;
+  bool _needsCompositingBitsUpdate = false; // set to true when a child is added
   /// Mark the compositing state for this render object as dirty.
   ///
-  /// When the subtree is mutated, we need to recompute our [needsCompositing]
-  /// bit, and our ancestors need to do the same (in case ours changed).
-  /// Therefore, [adoptChild] and [dropChild] call
-  /// [markNeedsCompositingBitsUpdate].
+  /// When the subtree is mutated, we need to recompute our
+  /// [needsCompositing] bit, and some of our ancestors need to do the
+  /// same (in case ours changed in a way that will change theirs). To
+  /// this end, [adoptChild] and [dropChild] call this method, and, as
+  /// necessary, this method calls the parent's, etc, walking up the
+  /// tree to mark all the nodes that need updating.
+  ///
+  /// This method does not schedule a rendering frame, because since
+  /// it cannot be the case that _only_ the compositing bits changed,
+  /// something else will have scheduled a frame for us.
   void _markNeedsCompositingBitsUpdate() {
     if (_needsCompositingBitsUpdate)
       return;
     _needsCompositingBitsUpdate = true;
-    final AbstractNode parent = this.parent;
-    if (parent is RenderObject)
-      parent._markNeedsCompositingBitsUpdate();
-    assert(parent == this.parent);
+    if (parent is RenderObject) {
+      final RenderObject parent = this.parent;
+      if (parent._needsCompositingBitsUpdate)
+        return;
+      if (!hasLayer && !parent.hasLayer) {
+        parent._markNeedsCompositingBitsUpdate();
+        return;
+      }
+    }
+    assert(() {
+      final AbstractNode parent = this.parent;
+      if (parent is RenderObject)
+        return parent._needsCompositing;
+      return true;
+    });
+    // parent is fine (or there isn't one), but we are dirty
+    _nodesNeedingCompositingBitsUpdate.add(this);
   }
 
-  bool _needsCompositing = false;
+  /// Updates the [needsCompositing] bits.
+  ///
+  /// Called as part of the rendering pipeline after [flushLayout] and before
+  /// [flushPaint].
+  static void flushCompositingBits() {
+    Timeline.startSync('Compositing Bits');
+    _nodesNeedingCompositingBitsUpdate.sort((RenderObject a, RenderObject b) => a.depth - b.depth);
+    for (RenderObject node in _nodesNeedingCompositingBitsUpdate) {
+      if (node.attached)
+        node._updateCompositingBits();
+    }
+    _nodesNeedingCompositingBitsUpdate.clear();
+    Timeline.finishSync();
```

Comment:
I considered that, but the code is all self-contained in object.dart, and I really don't see a way in which it could throw. Do you still want a try/finally block?